src/__tests__/poller.test.ts

```diff
@@ -0,0 +1,135 @@
+import Poller from '../poller'
+import type { Options } from '../types'
+import { getRandom } from '../utils'
+
+jest.useFakeTimers()
+
+jest.mock('../utils.ts', () => ({
+  getRandom: jest.fn()
+}))
+
+interface PollerArgs {
+  fetchFlags: jest.MockedFunction<() => Promise<any>>
+  configurations: Partial<Options>
+  pollInterval: number
+  maxAttempts: number
+}
+
+interface TestArgs {
+  delayFunction: jest.MockedFunction<() => number>
+  delayMs: number
+}
+
+const getPoller = (overrides: Partial<PollerArgs> = {}): { poller: Poller; pollerArgs: PollerArgs } => {
+  const args: PollerArgs = {
+    fetchFlags: jest.fn(),
+    configurations: {},
+    pollInterval: 60000,
+    maxAttempts: 5,
+    ...overrides
+  }
+
+  return {
+    poller: new Poller(args.fetchFlags, args.configurations, args.pollInterval),
+    pollerArgs: args
+  }
+}
+
+const getTestArgs = (overrides: Partial<TestArgs> = {}): TestArgs => {
+  const { delayMs = 5000 } = overrides // Extract delayMs or set a default of 5000
+
+  return {
+    delayMs,
+    delayFunction: (getRandom as jest.Mock).mockReturnValue(delayMs),
+    ...overrides
+  }
+}
+
+describe('Poller', () => {
+  it('should not start polling if it is already polling', () => {
+    const { poller } = getPoller()
+    const logSpy = jest.spyOn(poller as any, 'logDebug')
+
+    poller.start()
+    poller.start()
+
+    expect(logSpy).toHaveBeenCalledWith(
+      'Polling was requested but is already running - only one poller can run at a time.'
+    )
+  })
+
+  it('should retry fetching if there is an error', async () => {
+    const { poller, pollerArgs } = getPoller()
+    const fetchFlags = pollerArgs.fetchFlags
+    fetchFlags.mockResolvedValue(new Error('Fetch Error'))
+
+    const testArgs = getTestArgs()
+
+
+    const logSpy = jest.spyOn(poller as any, 'logDebug')
+
+    poller.start()
+
+    jest.advanceTimersByTime(pollerArgs.pollInterval)
+
+    // Allow first attempt to resolve
+    await Promise.resolve()
+
+    // Advance past the first delay
+    jest.advanceTimersByTime(testArgs.delayMs)
+
+    // Allow successful attempt to resolve
+    await Promise.resolve()
+
+    expect(fetchFlags).toHaveBeenCalledTimes(2)
+    expect(logSpy).toHaveBeenCalledWith('Error when polling for flag updates', expect.any(Error))
+  })
+
+  it('should not retry after max attempts are exceeded', async () => {
+    const { poller, pollerArgs } = getPoller()
+    const fetchFlags = pollerArgs.fetchFlags
+    const pollInterval = pollerArgs.pollInterval
+    const maxAttempts = pollerArgs.maxAttempts
+    fetchFlags.mockImplementation(() => {
+      return Promise.resolve(new Error('Fetch flags Error'))
+    })
```

Comment:
Again, you can flip the responsibilities here.
```suggestion
    const fetchFlagsMock = jest.fn().mockResolvedValues(new Error('Fetch flags Error'))
    const pollInterval = 100
    
    const { poller } = getPoller({ fetchFlags: fetchFlagsMock, pollInterval })
```

Reply:
Thanks - can't believe I went to all that trouble just to return the args. Pattern noted in toolbox.

Reply:
This actually isn't working - `fetchFlagsMock` is not itself being called. (0 times)

` it('should retry fetching if there is an error', async () => {
    let error = new Error('Fetch Error')
    const pollInterval = 100
    const fetchFlagsMock = jest.fn().mockResolvedValue(error)

    const { poller } = getPoller({fetchFlags: fetchFlagsMock, pollInterval: pollInterval})


    const testArgs = getTestArgs()

    poller.start()

    jest.advanceTimersByTime(pollInterval)

    // Allow first attempt to resolve
    await Promise.resolve()

    // Advance past the first delay
    jest.advanceTimersByTime(testArgs.delayMs)

    // Allow successful attempt to resolve
    await Promise.resolve()

    expect(fetchFlagsMock).toHaveBeenCalledTimes(2)
    expect(testArgs.logSpy).toHaveBeenCalled()
  })`

Reply:
Sorry @knagurski it's working perfectly - the poll interval was `100` which I coped from your example, but we needed to fast forward by 60000

Reply:
Just pulled it down and checked and the `0` is genuine. If you grab `pollerArgs.fetchFlags` out and assert on that, it'll give you the same result. Something changed in the implementation that's causing it.